## Normalize the data

In [1]:
import pandas as pd
import numpy as np

To download and read dataset into dataframe

In [2]:
!wget https://cocl.us/concrete_data
!mv concrete_data concrete_data.csv
data=pd.read_csv('concrete_data.csv')
data.head()

--2020-11-03 10:07:05--  https://cocl.us/concrete_data
Resolving cocl.us (cocl.us)... 169.63.96.194, 169.63.96.176
Connecting to cocl.us (cocl.us)|169.63.96.194|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv [following]
--2020-11-03 10:07:06--  https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv
Resolving s3-api.us-geo.objectstorage.softlayer.net (s3-api.us-geo.objectstorage.softlayer.net)... 67.228.254.196
Connecting to s3-api.us-geo.objectstorage.softlayer.net (s3-api.us-geo.objectstorage.softlayer.net)|67.228.254.196|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 58988 (58K) [text/csv]
Saving to: ‘concrete_data’

concrete_data       100%[===================>]  57.61K  --.-KB/s    in 0.1s    

2020-11-03 10:07:07 (388 KB/s) - ‘concre

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

#### Let's check how many data points we have.

In [3]:
data.shape

(1030, 9)

So, there are approximately 1030 samples to train model on.

To check dataset for missing values:

In [4]:
data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [5]:
data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

So data looks very clean and is ready to be used to build model.

#### Split data into predictors and target

The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.

In [6]:
concrete_data_columns=data.columns
predictors=data[concrete_data_columns[concrete_data_columns != 'Strength']] # All columns except Strength
target=data['Strength'] # Strength column

Let's do a quick sanity check of the predictors and the target dataframes.

In [7]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [8]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

### To normalize data by subtracting mean from individual predictors and dividing by standard deviation:

In [10]:
pnorm=(predictors-predictors.mean())/predictors.std()
pnorm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [13]:
ncol=pnorm.shape[1] # Total predictors
ncol

8

## Import Keras

In [15]:
import keras

In [16]:
from keras.models import Sequential
from keras.layers import Dense

To create model that has one hidden layer with 10 nodes and a ReLU activation function. It uses the adam optimizer and mean squared error as the loss function.

In [18]:
# define regression model
def regression_model():
    model = Sequential()
    model.add(Dense(10,activation='relu',input_shape=(ncol,)))
    model.add(Dense(1))
    model.compile(optimizer='adam',loss='mean_squared_error')
    return model

In [17]:
from sklearn.model_selection import train_test_split

Randomly split data into training and test sets by holding 30% of data for testing

In [19]:
X_train, X_test, y_train, y_test=train_test_split(pnorm, target, test_size=0.3, random_state=42)

## Train and Test the Network

Let's call the function now to create our model.

In [20]:
model=regression_model()

To train model on training data using 50 epochs:

In [21]:
epochs=50
model.fit(X_train,y_train,epochs=epochs,verbose=1)

Epoch 1/50
23/23 [==============================] - 0s 2ms/step - loss: 1587.1951
Epoch 2/50
23/23 [==============================] - 0s 3ms/step - loss: 1572.8613
Epoch 3/50
23/23 [==============================] - 0s 3ms/step - loss: 1558.4583
Epoch 4/50
23/23 [==============================] - 0s 2ms/step - loss: 1543.8689
Epoch 5/50
23/23 [==============================] - 0s 3ms/step - loss: 1528.7290
Epoch 6/50
23/23 [==============================] - 0s 2ms/step - loss: 1513.0184
Epoch 7/50
23/23 [==============================] - 0s 2ms/step - loss: 1496.5280
Epoch 8/50
23/23 [==============================] - 0s 2ms/step - loss: 1479.3794
Epoch 9/50
23/23 [==============================] - 0s 2ms/step - loss: 1461.2183
Epoch 10/50
23/23 [==============================] - 0s 2ms/step - loss: 1441.7942
Epoch 11/50
23/23 [==============================] - 0s 2ms/step - loss: 1421.3387
Epoch 12/50
23/23 [==============================] - 0s 2ms/step - loss: 1399.7053
Epoch 13/50
2

Next we need to evaluate the model on the test data.

In [22]:
loss_val=model.evaluate(X_test,y_test)
ypred=model.predict(X_test)
loss_val

10/10 [==============================] - 0s 2ms/step - loss: 315.8492


315.8492126464844

Now we need to compute the mean squared error between the predicted concrete strength and the actual concrete strength.

In [23]:
from sklearn.metrics import mean_squared_error

In [25]:
mse=mean_squared_error(y_test, ypred)
mean = np.mean(mse)
sd=np.std(mse)
print(mean,sd)

315.8491913624639 0.0


Create a list of 50 mean squared errors and report mean and the standard deviation of the mean squared errors.

In [27]:
totalmse=50
epochs=50
mean_squared_errors=[]
for i in range(0, totalmse):
    X_train, X_test, y_train, y_test=train_test_split(pnorm,target,test_size=0.3,random_state=i)
    model.fit(X_train, y_train,epochs=epochs,verbose=0)
    MSE = model.evaluate(X_test,y_test,verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mse)
mean_squared_errors=np.array(mean_squared_errors)
mean=np.mean(mean_squared_errors)
sd=np.std(mean_squared_errors)
print("\nMean and standard deviation of " +str(totalmse) + " mean squared errors using normalized data:\n")
print("Mean: "+str(mean))
print("Standard Deviation: "+str(sd))

MSE 1: 36.57171630859375
MSE 2: 41.61995315551758
MSE 3: 33.24441146850586
MSE 4: 38.24230194091797
MSE 5: 38.982765197753906
MSE 6: 39.35704040527344
MSE 7: 44.460723876953125
MSE 8: 35.951820373535156
MSE 9: 36.528438568115234
MSE 10: 35.67123031616211
MSE 11: 36.04814910888672
MSE 12: 35.044925689697266
MSE 13: 40.8748893737793
MSE 14: 42.203983306884766
MSE 15: 34.857662200927734
MSE 16: 33.507484436035156
MSE 17: 36.962894439697266
MSE 18: 37.11610794067383
MSE 19: 37.06245040893555
MSE 20: 36.802284240722656
MSE 21: 35.0628776550293
MSE 22: 36.29821014404297
MSE 23: 32.54432678222656
MSE 24: 37.856788635253906
MSE 25: 36.73625946044922
MSE 26: 38.84296798706055
MSE 27: 33.64200210571289
MSE 28: 31.595661163330078
MSE 29: 38.68289566040039
MSE 30: 37.13102722167969
MSE 31: 35.12486267089844
MSE 32: 32.94158172607422
MSE 33: 34.44196701049805
MSE 34: 36.251548767089844
MSE 35: 36.86576843261719
MSE 36: 41.214263916015625
MSE 37: 36.52311325073242
MSE 38: 39.73874282836914
MSE 39: 3